vm/src/dictdatatype.rs

```diff
@@ -550,22 +540,34 @@ impl<T: Clone> Dict<T> {
                 });
                 loop {
                     let index_index = idxs.next();
-                    match IndexEntry::from(inner.indices[index_index]) {
-                        IndexEntry::Dummy => {
-                            if freeslot.is_none() {
-                                freeslot = Some(index_index);
+                    let index_entry = *unsafe { inner.indices.get_unchecked(index_index) };
+                    match index_entry {
+                        IndexEntry::DUMMY => {
+                            if free_slot.is_none() {
+                                free_slot = Some(index_index);
                             }
                         }
-                        IndexEntry::Free => {
-                            let idxs = match freeslot {
-                                Some(free) => (IndexEntry::Dummy, free),
-                                None => (IndexEntry::Free, index_index),
+                        IndexEntry::FREE => {
+                            let idxs = match free_slot {
+                                Some(free) => (IndexEntry::DUMMY, free),
+                                None => (IndexEntry::FREE, index_index),
                             };
                             return Ok(idxs);
                         }
-                        IndexEntry::Index(i) => {
-                            let entry = &inner.entries[i].as_ref().unwrap();
-                            let ret = (IndexEntry::Index(i), index_index);
+                        idx => {
+                            let entry = unsafe {
+                                // DUMMY and FREE are already handled above.
+                                // i is always valid and entry always exists.
+                                let i = idx
+                                    .index()
+                                    .unwrap_or_else(|| std::hint::unreachable_unchecked());
+                                inner
+                                    .entries
+                                    .get_unchecked(i)
+                                    .as_ref()
+                                    .unwrap_or_else(|| std::hint::unreachable_unchecked())
```

Comment:
Honestly, I'd rather just have just `idx.0` or `idx.as_i64()` or something, the risk (which I know, this is totally fine) of unreachable_unchecked UB seems totally unnecessary here. alternatively, `index()` could return a `Result<usize, bool>` ? basically the same tristate enum as before but passed around less, I guess.

Reply:
unreachable_unchecked is just same unsafe as `idx.0 as usize`. But `idx.0 as usize` hides it can be unsafe value without proper preprocess, while `.index()` shows we are doing unsafe operation by trusting the preprocess. Even more, `unreachable_unchecked` doesn't work without debug mode. This is safer than naive casting.

Reply:
Well, you can at least use `option.unwrap_unchecked()` I guess.

Reply:
Oh, I didn't know that's stablized